python/ray/serve/_private/application_state.py

```diff
@@ -421,18 +422,27 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
         if self._target_state.deleting:
             return ApplicationStatus.DELETING, ""
 
-        num_healthy_deployments = 0
+        num_healthy_or_autoscaling_deployment = 0
+        num_updating_deployments = 0
         unhealthy_deployment_names = []
 
         for deployment_status in self.get_deployments_statuses():
             if deployment_status.status == DeploymentStatus.UNHEALTHY:
                 unhealthy_deployment_names.append(deployment_status.name)
-            if deployment_status.status == DeploymentStatus.HEALTHY:
-                num_healthy_deployments += 1
+            elif (
+                deployment_status.status == DeploymentStatus.HEALTHY
+                or deployment_status.status_trigger == DeploymentStatusTrigger.AUTOSCALE
+            ):
+                # If all deployments are HEALTHY or autoscaling, then
+                # application is RUNNING
+                num_healthy_or_autoscaling_deployment += 1
+            else:
+                # If deployments are UPDATING or UPSCALING/DOWNSCALING
+                # with status trigger CONFIG_UPDATE, then application is
+                # still DEPLOYING
+                num_updating_deployments += 1
```

Comment:
Thought a little bit about it, I think when the new config is applied, the deployment status will be `UPDATING`. Then in the first control loop iteration that follows that, the status will switch to `UPSCALING` because `autoscale()` will adjust the target num replicas to something within the new [min,max] range.

This also made me realize a bigger problem which is if the deployment is in the middle of updating for whatever reason (even if its a heavyweight change), `autoscale()` will change the status to `UPSCALING`. I can probably change the logic to set status to `UPSCALING/DOWNSCALING` only if the current status is already `HEALTHY`.

Reply:
For maintaining all these status change, i think we can add update_trigger_status in the central place (maybe in DeploymentStatusInfo).
So that we can list out all possible combinations (as the graph shows), and raise Error if we mistakenly update some status.

Reply:
@zcin that sounds good. You'll also need to handle the case where you are `UPDATING` and still scaling up, but the replicas are within the bounds of `[min_replicas, max_replicas]`. In that case, the status should flip to `UPSCALING` (or `DOWNSCALING`).

I also like Sihan's suggestion.

Reply:
Can you clarify what `update_trigger_status` would do? Right now `UPDATING`, `UPSCALING`, `DOWNSCALING` are set in a centralized place (`set_target_state`), and `HEALTHY` and `UNHEALTHY` must be set in other parts of the deployment state manager just be nature of how we check for them.

In the meantime, I've fixed the behavior that @edoakes brought up and added a test for it.

Reply:
@edoakes That's a good point, although I think that's not super straightforward to fix because it's hard to know whether you are still `UPDATING` because of something else (e.g. code version change or `user_config` change) or if you're still `UPDATING` simply because you haven't reached [min, max replicas] yet. Usually I would expect the status to go from `UPDATING` -> `HEALTHY` -> `UPSCALING/DOWNSCALING`, so I'd consider this an edge case. I've created an issue to track it: https://github.com/ray-project/ray/issues/41038

Reply:
@sihanwang41 Can you clarify what `update_trigger_status` would do? Right now `UPDATING`, `UPSCALING`, `DOWNSCALING` are set in a centralized place (`set_target_state`), and `HEALTHY` and `UNHEALTHY` must be set in other parts of the deployment state manager because we check for them in different methods.

In the meantime, I've fixed the behavior that @edoakes brought up and added a test for it.

Reply:
oh, we can explicitly list all possible combinations to avoid setting up unexpected status. (e.g.  deployment status "UPSCALING" + trigger status "UPSCALE_COMPLETED")

Reply:
I see, that makes sense! I spent some time prototyping something that is more centralized and it seems promising, but would prefer to add it to a follow up PR since it changes some existing implementation details.